File: frontend/src/api/exchangeHardening.ts
import { apiClient } from '@/api/client';

// --- Types ---

export interface PassportStateDiff {
  id: string;
  publication_id: string;
  prior_publication_id: string | null;
  diff_summary: {
    added_sections: string[];
    removed_sections: string[];
    changed_sections: { section: string; field: string; old: string | null; new: string | null }[];
  } | null;
  sections_changed_count: number;
  computed_at: string | null;
  created_at: string;
}

export interface ExchangeValidationReport {
  id: string;
  import_receipt_id: string;
  schema_valid: boolean | null;
  contract_valid: boolean | null;
  version_valid: boolean | null;
  hash_valid: boolean | null;
  identity_safe: boolean | null;
  validation_errors: { check: string; message: string; severity: string }[] | null;
  overall_status: string;
  validated_at: string | null;
  validated_by_user_id: string | null;
  created_at: string;
}

export interface ContributorRequest {
  id: string;
  building_id: string;
  contributor_type: string;
  scope_description: string | null;
  access_token: string;
  expires_at: string;
  status: string;
  created_by_user_id: string;
  created_at: string;
  updated_at: string;
}

export interface ContributorSubmission {
  id: string;
  request_id: string;
  contributor_org_id: string | null;
  contributor_name: string | null;
  submission_type: string;
  file_url: string | null;
  structured_data: Record<string, unknown> | null;
  notes: string | null;
  status: string;
  reviewed_by_user_id: string | null;
  reviewed_at: string | null;
  review_notes: string | null;
  created_at: string;
}

export interface ContributorReceipt {
  id: string;
  submission_id: string;
  document_id: string | null;
  receipt_hash: string;
  accepted_at: string;
  created_at: string;
}

export interface WebhookSubscription {
  id: string;
  partner_org_id: string;
  endpoint_url: string;
  subscribed_events: string[] | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

// --- API ---

export const exchangeHardeningApi = {
  async getPublicationDiff(publicationId: string): Promise<PassportStateDiff> {
    const res = await apiClient.get<PassportStateDiff>(`/publications/${publicationId}/diff`);
    return res.data;
  },

  async validateImport(receiptId: string): Promise<ExchangeValidationReport> {
    const res = await apiClient.post<ExchangeValidationReport>(
      `/import-receipts/${receiptId}/validate`,
    );
    return res.data;
  },

  async reviewImport(receiptId: string, decision: string): Promise<unknown> {
    const res = await apiClient.post(`/import-receipts/${receiptId}/review`, { decision });
    return res.data;
  },

  async integrateImport(receiptId: string): Promise<unknown> {
    const res = await apiClient.post(`/import-receipts/${receiptId}/integrate`);
    return res.data;
  },

  async listContributorRequests(buildingId?: string): Promise<ContributorRequest[]> {
    const params = buildingId ? { building_id: buildingId } : {};
    const res = await apiClient.get<ContributorRequest[]>('/contributor-requests', { params });
    return res.data;
  },

  async createContributorRequest(data: {
    building_id: string;
    contributor_type: string;
    scope_description?: string;
  }): Promise<ContributorRequest> {
    const res = await apiClient.post<ContributorRequest>('/contributor-requests', data);
    return res.data;
  },

  async listPendingSubmissions(): Promise<ContributorSubmission[]> {
    const res = await apiClient.get<ContributorSubmission[]>('/contributor-submissions/pending');
    return res.data;
  },

  async acceptSubmission(submissionId: string): Promise<ContributorReceipt> {
    const res = await apiClient.post<ContributorReceipt>(
      `/contributor-submissions/${submissionId}/accept`,
    );
    return res.data;
  },

  async rejectSubmission(submissionId: string, notes?: string): Promise<ContributorSubmission> {
    const res = await apiClient.post<ContributorSubmission>(
      `/contributor-submissions/${submissionId}/reject`,
      { notes },
    );
    return res.data;
  },
};

File: backend/app/schemas/exchange_hardening.py
"""BatiConnect — Exchange Hardening + Contributor Gateway schemas."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

# --- PassportStateDiff ---


class DiffSectionChange(BaseModel):
    section: str
    field: str
    old: str | None = None
    new: str | None = None


class DiffSummary(BaseModel):
    added_sections: list[str] = []
    removed_sections: list[str] = []
    changed_sections: list[DiffSectionChange] = []


class PassportStateDiffRead(BaseModel):
    id: UUID
    publication_id: UUID
    prior_publication_id: UUID | None
    diff_summary: dict | None
    sections_changed_count: int
    computed_at: datetime | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- ExchangeValidationReport ---


class ValidationError(BaseModel):
    check: str
    message: str
    severity: str = "error"


class ExchangeValidationReportRead(BaseModel):
    id: UUID
    import_receipt_id: UUID
    schema_valid: bool | None
    contract_valid: bool | None
    version_valid: bool | None
    hash_valid: bool | None
    identity_safe: bool | None
    validation_errors: list[dict] | None
    overall_status: str
    validated_at: datetime | None
    validated_by_user_id: UUID | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ImportReviewRequest(BaseModel):
    decision: str  # passed|failed|review_required


# --- ExternalRelianceSignal ---


class RelianceSignalCreate(BaseModel):
    publication_id: UUID | None = None
    import_receipt_id: UUID | None = None
    signal_type: str  # consumed|acknowledged|superseded|disputed
    partner_org_id: UUID | None = None
    notes: str | None = None


class RelianceSignalRead(BaseModel):
    id: UUID
    publication_id: UUID | None
    import_receipt_id: UUID | None
    signal_type: str
    partner_org_id: UUID | None
    notes: str | None
    recorded_at: datetime
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- PartnerWebhookSubscription ---


class WebhookSubscriptionCreate(BaseModel):
    partner_org_id: UUID
    endpoint_url: str
    hmac_secret: str
    subscribed_events: list[str] = []
    is_active: bool = True


class WebhookSubscriptionRead(BaseModel):
    id: UUID
    partner_org_id: UUID
    endpoint_url: str
    subscribed_events: list[str] | None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class DeliveryAttemptRead(BaseModel):
    id: UUID
    subscription_id: UUID
    event_type: str
    idempotency_key: str
    payload: dict | None
    status: str
    http_status: int | None
    error_message: str | None
    attempt_count: int
    last_attempt_at: datetime | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- ContributorGatewayRequest ---


class ContributorRequestCreate(BaseModel):
    building_id: UUID
    contributor_type: str  # contractor|lab
    scope_description: str | None = None
    expires_in_hours: int = 72
    linked_procedure_id: UUID | None = None
    linked_remediation_id: UUID | None = None


class ContributorRequestRead(BaseModel):
    id: UUID
    building_id: UUID
    contributor_type: str
    scope_description: str | None
    access_token: str
    expires_at: datetime
    status: str
    created_by_user_id: UUID
    linked_procedure_id: UUID | None
    linked_remediation_id: UUID | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- ContributorSubmission ---


class ContributorSubmissionCreate(BaseModel):
    contributor_org_id: UUID | None = None
    contributor_name: str | None = None
    submission_type: str  # completion_report|lab_results|certificate|attestation|photo_evidence|other
    file_url: str | None = None
    structured_data: dict | None = None
    notes: str | None = None


class ContributorSubmissionRead(BaseModel):
    id: UUID
    request_id: UUID
    contributor_org_id: UUID | None
    contributor_name: str | None
    submission_type: str
    file_url: str | None
    structured_data: dict | None
    notes: str | None
    status: str
    reviewed_by_user_id: UUID | None
    reviewed_at: datetime | None
    review_notes: str | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class SubmissionRejectRequest(BaseModel):
    notes: str | None = None


# --- ContributorReceipt ---


class ContributorReceiptRead(BaseModel):
    id: UUID
    submission_id: UUID
    document_id: UUID | None
    evidence_link_id: UUID | None
    proof_delivery_id: UUID | None
    receipt_hash: str
    accepted_at: datetime
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)
